### PR TITLE
refactor(core): Refactor parts of event_replay into a shared library and move stashing function to listening code

### PR DESCRIPTION
**Pull request by iteriani**
This also moves the code that stashes the jsaction more closely to the code that actually sets the event listener. I've considered just moving that code to the listener itself..but it's in a different package. Maybe we need to add another DomEventsPlugin implementation?

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/hydration/annotate.ts [hunk 1]
```diff
@@ -432,10 +432,15 @@ function serializeLView(lView: LView, context: HydrationContext): SerializedView
       continue;
     }
 
+    // Attach `jsaction` attribute to elements that have registered listeners,
+    // thus potentially having a need to do an event replay.
     if (nativeElementsToEventTypes) {
-      // Attach `jsaction` attribute to elements that have registered listeners,
-      // thus potentially having a need to do an event replay.
-      setJSActionAttribute(tNode, lView[i], nativeElementsToEventTypes);
+      if (tNode.type & TNodeType.Element) {
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/annotate.ts:438]: nit: we can merge this `if` with the previous one.

File: packages/core/src/hydration/event_replay.ts [hunk 2]
```diff
@@ -62,24 +57,37 @@ export function withEventReplay(): Provider[] {
       provide: IS_EVENT_REPLAY_ENABLED,
       useValue: true,
     },
+    {
+      provide: EVENT_CONTRACT_FROM_SSR,
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_replay.ts:61]: If we plan to reuse this token later for client-only cases (do we plan to check if this token is provided and if not - create a new event contract?), it'd be be great to abstract out the location where the event contract was created from the actual code that would use it. Potentially, a singleton service (that would act as a container) might be a good approach, for example:

```
@Injectable({providedIn: 'root'})
export class GlobalEventDelegation {
  contract: EventContract|null = null;
  dispatcher: Dispatcher|null = null;
}
```

In this case, in the `APP_BOOTSTRAP_LISTENER` below, you inject it (it'll be instantiated if it's the first time we inject it):

```
const eventDelegation = injector.get(GlobalEventDelegation);
eventDelegation.contract = ...;
eventDelegation.dispatcher = ...;
```

Later on in the code that adds event listeners, we can also inject a class and work with the `contract` and `dispatcher` fields.

File: packages/core/src/hydration/event_replay.ts [hunk 3]
```diff
@@ -62,24 +57,37 @@ export function withEventReplay(): Provider[] {
...
-              jsactionSet.add(el);
-              el.__jsaction_fns = new Map();
-            }
-            const eventMap = el.__jsaction_fns!;
-            if (!eventMap.has(eventName)) {
-              eventMap.set(eventName, []);
-            }
-            eventMap.get(eventName)!.push(listenerFn);
-          }
+        const injector = inject(Injector);
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_replay.ts:87]: We don't seem to use this const.

File: packages/core/src/hydration/event_shared.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/**
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_shared.ts:1]: I'd propose moving this shared logic outside of the `hydration` folder (to avoid confusion on why certain functions are used in the client-only code), for example to the `packages/core/src/event_delegation_utils.ts`

File: packages/core/src/hydration/event_shared.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {EarlyJsactionDataContainer} from '@angular/core/primitives/event-dispatch';
+import {JSACTION} from '@angular/core/primitives/event-dispatch/src/attribute';
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_shared.ts:10]: Do we have the `JSACTION` symbol in the public API of the primitive (can we move it to the previous import)?

File: packages/core/src/hydration/event_shared.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import {RElement} from '../render3/interfaces/renderer_dom';
+
+const JSACTION_ATTRIBUTE = 'jsaction';
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_shared.ts:13]: This should probably be coming from the `event-dispatch` lib, but we can refactor later.

File: packages/core/src/hydration/event_replay.ts [hunk 7]
```diff
@@ -95,33 +103,16 @@ export function withEventReplay(): Provider[] {
...
-                }
-                // After hydration, we shouldn't need to do anymore work related to
-                // event replay anymore.
-                setDisableEventReplayImpl(() => {});
+              const eventContract = injector.get(EVENT_CONTRACT_FROM_SSR)!;
+              const dispatcher = new EventDispatcher(handleEvent);
+              registerDispatcher(eventContract, dispatcher);
+              for (const el of jsactionSet) {
+                el.removeAttribute(JSACTION_ATTRIBUTE);
+                el.__jsaction_fns = undefined;
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_replay.ts:111]: I think it'd be great to isolate all `__jsaction_fns` reads and writes in the newly-created utils file, i.e. introduce a couple functions like `getListeners` and `removeListeners` and use those functions in other places. That can help with readability and we can also add docs to those helpers to improve the DX.

File: packages/core/src/hydration/annotate.ts (hunk 1 above)
Comment (iteriani) [reply packages/core/src/hydration/annotate.ts]: done

File: packages/core/src/hydration/event_replay.ts (hunk 2 above)
Comment (iteriani) [reply packages/core/src/hydration/event_replay.ts]: done

File: packages/core/src/hydration/event_shared.ts (hunk 4 above)
Comment (iteriani) [reply packages/core/src/hydration/event_shared.ts]: done

File: packages/core/src/hydration/event_shared.ts (hunk 6 above)
Comment (iteriani) [reply packages/core/src/hydration/event_shared.ts]: done

File: packages/core/src/hydration/event_replay.ts (hunk 7 above)
Comment (iteriani) [reply packages/core/src/hydration/event_replay.ts]: done

File: packages/core/src/hydration/event_shared.ts (hunk 5 above)
Comment (iteriani) [reply packages/core/src/hydration/event_shared.ts]: done

File: packages/core/src/event_delegation_utils.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  if (!handlerFns) {
+    return;
+  }
+  for (const handler of handlerFns) {
+    handler(event);
+  }
+}
+
+export function setJSActionAttribute(nativeElement: Element, eventTypes: string[]) {
+  const parts = eventTypes.map((event) => `${event}:`);
```
Comment (tbondwilkinson) [inline packages/core/src/event_delegation_utils.ts:40]: `eventTypes.map((eventType) => `${eventType}:`)`

File: packages/core/src/event_delegation_utils.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  if (parts.length > 0) {
```
Comment (tbondwilkinson) [inline packages/core/src/event_delegation_utils.ts:41]: Move this to an early return at the start of the function.

`if (!eventTypes.length) return;`

File: packages/core/src/event_delegation_utils.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    nativeElement.setAttribute(
+      JSACTION,
+      `${nativeElement.getAttribute(JSACTION) ?? ''}${parts.join(';')}`,
+    );
+  }
+}
+
+export const sharedStashFunction = (el: Element, eventName: string, listenerFn: VoidFunction) => {
```
Comment (tbondwilkinson) [inline packages/core/src/event_delegation_utils.ts:49]: Why VoidFunction instead of () => void

File: packages/core/src/event_delegation_utils.ts [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  const eventMap = el.__jsaction_fns || (el.__jsaction_fns = new Map());
+  if (!eventMap.has(eventName)) {
+    eventMap.set(eventName, []);
+  }
+  eventMap.get(eventName)!.push(listenerFn);
```
Comment (tbondwilkinson) [inline packages/core/src/event_delegation_utils.ts:54]:
```suggestion
  const eventListenerMap = el.__jsaction_fns ?? new Map();
  const eventListeners = eventListenerMap.get(eventType) ?? [];
  eventListeners.push(listenerFn);
  eventListenerMap.set(eventType, eventListeners);
  el.__jsaction_fns = eventListenerMap;
```

**Review by tbondwilkinson — APPROVED** [review]

File: packages/core/primitives/event-dispatch/index.ts [hunk 12]
```diff
@@ -15,3 +15,4 @@ export {bootstrapEarlyEventContract} from './src/register_events';
 export type {EventContractTracker} from './src/register_events';
 export {EventInfoWrapper} from './src/event_info';
 export {isSupportedEvent, isCaptureEvent} from './src/event_type';
+export * from './src/attribute';
```
Comment (AndrewKushnir) [inline packages/core/primitives/event-dispatch/index.ts:18]: nit: it should probably be in a separate PR, so that we can land primitives changes separately from FW?

File: packages/core/src/event_delegation_utils.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    handler(event);
+  }
+}
+
+export function setJSActionAttribute(nativeElement: Element, eventTypes: string[]) {
+  const parts = eventTypes.map((event) => `${event}:`);
+  if (parts.length > 0) {
+    // to appease the security test
+    nativeElement.setAttribute(
+      'jsaction',
```
Comment (AndrewKushnir) [inline packages/core/src/event_delegation_utils.ts:44]:
```suggestion
      JSACTION,
```

File: packages/core/src/event_delegation_utils.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+declare global {
+  var ngContracts: {[key: string]: EarlyJsactionDataContainer};
+  interface Element {
+    __jsaction_fns: Map<string, Function[]> | undefined;
+  }
+}
+
+export const CONTRACT_PROPERTY = 'ngContracts';
+
+export function handleEvent(event: Event) {
```
Comment (AndrewKushnir) [inline packages/core/src/event_delegation_utils.ts:29]: nit: we can give this function a bit more descriptive name, for example:
```suggestion
export function invokeRegisteredListeners(event: Event) {
```

File: packages/core/src/event_delegation_utils.ts [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    for (const et of earlyJsactionData.etc) {
+      eventDelegation.eventContract.addEvent(et);
+    }
+    eventDelegation.eventContract.replayEarlyEvents(container);
+  } else {
+    eventDelegation.eventContract = new EventContract(
+      new EventContractContainer(globalThis.document.body),
+    );
+  }
```
Comment (AndrewKushnir) [inline packages/core/src/event_delegation_utils.ts:105]: I think we'd need to separate these 2 branches into 2 functions:
- one function would init event contract based on SSR'ed content
- another function would init event contract for client-only cases

The first function would be invoked when `withEventReplay()` is included, the second function would be invoked in a new public API function (e.g. `provideGlobalEventDelegation()`) if event contract isn't set up already. This should allow us to tree-shake-away SSR-aware part.

File: packages/core/src/hydration/annotate.ts [hunk 16]
```diff
@@ -432,10 +429,13 @@ function serializeLView(lView: LView, context: HydrationContext): SerializedView
       continue;
     }
 
-    if (nativeElementsToEventTypes) {
-      // Attach `jsaction` attribute to elements that have registered listeners,
-      // thus potentially having a need to do an event replay.
-      setJSActionAttribute(tNode, lView[i], nativeElementsToEventTypes);
+    // Attach `jsaction` attribute to elements that have registered listeners,
+    // thus potentially having a need to do an event replay.
+    if (nativeElementsToEventTypes && tNode.type & TNodeType.Element) {
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/annotate.ts:434]: nit (mostly for readability):
```suggestion
    if (nativeElementsToEventTypes && (tNode.type & TNodeType.Element)) {
```

File: packages/core/primitives/event-dispatch/index.ts (hunk 12 above)
Comment (thePunderWoman) [reply packages/core/primitives/event-dispatch/index.ts:18]: I think this should be fine to be in this PR since they are related. PRs don't have to be exclusive in that regard. The main issue is rollback safety, which shouldn't matter in this case.

File: packages/core/src/hydration/event_replay.ts [hunk 17]
```diff
@@ -180,28 +135,3 @@ export function collectDomEventsInfo(
...
-    const nativeElement = unwrapRNode(rNode) as Element;
-    const events = nativeElementToEvents.get(nativeElement) ?? [];
-    const parts = events.map((event) => `${event}:`);
-    if (parts.length > 0) {
-      nativeElement.setAttribute(JSACTION_ATTRIBUTE, parts.join(';'));
-    }
-  }
-}
-
-function handleEvent(event: Event) {
```
Comment (thePunderWoman) [inline packages/core/src/hydration/event_replay.ts:199]: I don't know if moving this to a util file makes the most sense since it's a core feature of event replay. Plus it's definitely modified in partial hydration cases. WDYT?

File: packages/core/src/event_delegation_utils.ts (hunk 11 above)
Comment (thePunderWoman) [reply packages/core/src/event_delegation_utils.ts:55]: We also do a lot of modifying of this method for partial hydration cases. Might be worth looking at it before we land this.

File: packages/core/src/event_delegation_utils.ts (hunk 15 above)
Comment (thePunderWoman) [reply packages/core/src/event_delegation_utils.ts:105]: +1 to @AndrewKushnir's suggestion

File: packages/core/src/hydration/annotate.ts (hunk 16 above)
Comment (iteriani) [reply packages/core/src/hydration/annotate.ts:434]: done

File: packages/core/src/event_delegation_utils.ts (hunk 15 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done

File: packages/core/src/event_delegation_utils.ts (hunk 13 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: there's a security test that fails when i do this :(

File: packages/core/src/event_delegation_utils.ts (hunk 11 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: tomw: done
@thePunderWoman wdym?

File: packages/core/src/event_delegation_utils.ts (hunk 10 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: no good reason

File: packages/core/src/event_delegation_utils.ts (hunk 9 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done

File: packages/core/src/event_delegation_utils.ts (hunk 8 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done-ish

File: packages/core/src/event_delegation_utils.ts (hunk 14 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done

File: packages/core/src/hydration/annotate.ts (hunk 16 above)
Comment (iteriani) [reply packages/core/src/hydration/annotate.ts:434]: done

File: packages/core/src/event_delegation_utils.ts (hunk 13 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done

File: packages/core/src/hydration/annotate.ts (hunk 16 above)
Comment (AndrewKushnir) [reply packages/core/src/hydration/annotate.ts:434]: Reopening this thread, it looks like there was no change made to this part of the code.

File: packages/core/src/event_delegation_utils.ts [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+  );
+  for (const et of earlyJsactionData.et) {
+    eventDelegation.eventContract.addEvent(et);
+  }
+  for (const et of earlyJsactionData.etc) {
+    eventDelegation.eventContract.addEvent(et);
+  }
+  eventDelegation.eventContract.replayEarlyEvents(container);
+  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
+  registerDispatcher(eventDelegation.eventContract, dispatcher);
```
Comment (AndrewKushnir) [inline packages/core/src/event_delegation_utils.ts:101]: nit: for better compression (since we don't rename properties), we can create a const and use it like this:
```suggestion
  const eventContract = new EventContract(
    new EventContractContainer(earlyJsactionData.c),
  );
  for (const et of earlyJsactionData.et) {
    eventContract.addEvent(et);
  }
  for (const et of earlyJsactionData.etc) {
    eventContract.addEvent(et);
  }
  eventContract.replayEarlyEvents(container);
  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
  registerDispatcher(eventContract, dispatcher);
  eventDelegation.eventContract = eventContract;
```

File: packages/core/src/event_delegation_utils.ts [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+export const initGlobalEventDelegation = (
+  eventDelegation: GlobalEventDelegation,
+  injector: Injector,
+) => {
+  if (injector.get(IS_EVENT_REPLAY_ENABLED, EVENT_REPLAY_ENABLED_DEFAULT)) {
+    return;
+  }
+  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
+  registerDispatcher(eventDelegation.eventContract, dispatcher);
+};
```
Comment (AndrewKushnir) [inline packages/core/src/event_delegation_utils.ts:113]: I think we can avoid referencing `IS_EVENT_REPLAY_ENABLED` here and instead check if the `eventDelegation.eventContract` is defined:
```suggestion
export const initGlobalEventDelegation = (
  eventDelegation: GlobalEventDelegation,
  eventContractContainer: Element,
) => {
  if (eventDelegation.eventContract) {
    // EventContract is already configured for this application,
    // no need to create a new one, exit.
    return;
  }
  const container = new EventContractContainer(eventContractContainer);
  const eventContract = new EventContract(container);
  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
  registerDispatcher(eventContract, dispatcher);
  eventDelegation.eventContract = eventContract;
};
```
I've also updated the function to init `EventContract` if needed. We can use this function later in the client-only code path. WDYT?

File: packages/core/src/event_delegation_utils.ts [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    if (this.eventContract) {
+      this.eventContract.addEvent(eventName);
+      setJSActionAttribute(el, [eventName]);
+      return true;
+    }
+    return false;
+  }
+}
+
+export const initEventReplay = (eventDelegation: GlobalEventDelegation, injector: Injector) => {
```
[the remark this answers is not in the document]
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done

File: packages/core/src/hydration/annotate.ts (hunk 16 above)
Comment (iteriani) [reply packages/core/src/hydration/annotate.ts:434]: i think prettier removes those parens

File: packages/core/src/event_delegation_utils.ts (hunk 19 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: i was a little afraid of order of operations in the case both eventReplay() and withGlobalEventDelegation() are available.  I did initialize the event contract though.

**Review by thePunderWoman — APPROVED** [review]
LGTM

I'm comfortable with these changes after some discussion with @iteriani.

reviewed-for: fw-core, fw-platform-server, public-api, primitives

File: packages/core/src/hydration/event_replay.ts (hunk 17 above)
Comment (iteriani) [reply packages/core/src/hydration/event_replay.ts:199]: discussed offline

File: packages/core/src/hydration/event_replay.ts [hunk 21]
```diff
@@ -95,33 +85,19 @@ export function withEventReplay(): Provider[] {
...
-                }
-                // After hydration, we shouldn't need to do anymore work related to
-                // event replay anymore.
-                setDisableEventReplayImpl(() => {});
+              const isGlobalEventDelegationEnabled = injector.get(
+                IS_GLOBAL_EVENT_DELEGATION_ENABLED,
+                false,
+              );
+              if (isGlobalEventDelegationEnabled) {
+                return;
```
Comment (AndrewKushnir) [inline packages/core/src/hydration/event_replay.ts:93]: nit: since we use this code in a couple places, we can extract it into a separate function and call it here:
```suggestion
              if (isGlobalEventDelegationEnabled(injector)) {
                return;
```
It'd also help to keep the default value the same in all locations.

**Review by AndrewKushnir — APPROVED** [review]
LGTM, just one comment with a refactoring proposal.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by AndrewKushnir — APPROVED** [review]

File: packages/core/src/hydration/event_replay.ts (hunk 21 above)
Comment (iteriani) [reply packages/core/src/hydration/event_replay.ts]: done

**Comment by iteriani** [comment]
TESTED=[TGP](https://fusion2.corp.google.com/presubmit/638459441/OCL:638459441:BASE:640238118:1717570178119:855a85d7/targets)

Failing tests appear unrelated

**Comment by thePunderWoman** [comment]
This PR was merged into the repository by commit 6e89ef1ebf19bed02832dc5d05480fdf95a7c929.
